### PR TITLE
Add VBV  support for VP9

### PR DESCRIPTION
**Pull request by kavithasampath**

### CODE REVIEW

**Comment by tianjunwork** [comment]
vbvBufInit and hrdFlag are not needed for vp9?

**Comment by karthikmcw** [comment]
> vbvBufInit and hrdFlag are not needed for vp9?

No, they are not required.

**Comment by tianjunwork** [comment]
appveyor shows build error on windows:
C:\projects\svt-vp9\Source\API\EbSvtVp9Enc.h(132): message :  see declaration of 'EbSvtVp9EncConfiguration' [C:\projects\svt-vp9\Source\Lib\Codec\SvtVp9Enc.vcxproj]
C:\projects\svt-vp9\Source\Lib\Codec\EbRateControlProcess.c(4238,88): error C2039:  'vbvBufsize': is not a member of 'EbSvtVp9EncConfiguration' [C:\projects\svt-vp9\Source\Lib\Codec\SvtVp9Enc.vcxproj]
C:\projects\svt-vp9\Source\API\EbSvtVp9Enc.h(132): message :  see declaration of 'EbSvtVp9EncConfiguration' [C:\projects\svt-vp9\Source\Lib\Codec\SvtVp9Enc.vcxproj]
C:\projects\svt-vp9\Source\Lib\Codec\EbRateControlProcess.c(4239,77): error C2039:  'vbvMaxrate': is not a member of 'EbSvtVp9EncConfiguration' [C:\projects\svt-vp9\Source\Lib\Codec\SvtVp9Enc.vcxproj]
C:\projects\svt-vp9\Source\API\EbSvtVp9Enc.h(132): message :  see declaration of 'EbSvtVp9EncConfiguration' [C:\projects\svt-vp9\Source\Lib\Codec\SvtVp9Enc.vcxproj]
C:\projects\svt-vp9\Source\Lib\Codec\EbRateControlProcess.c(4240,77): error C2039:  'vbvBufsize': is not a member of 'EbSvtVp9EncConfiguration' [C:\projects\svt-vp9\Source\Lib\Codec\SvtVp9Enc.vcxproj]

**Comment by tianjunwork** [comment]
@kavithasampath Thank you addressing my comments. Could you add the sign-off line in your commit message?

File: Source/Lib/Codec/EbRateControlProcess.c [hunk 1]
```diff
@@ -4338,7 +4483,12 @@ void* rate_control_kernel(void *input_ptr)
 
                 picture_control_set_ptr->parent_pcs_ptr->cpi->common.base_qindex = picture_control_set_ptr->base_qindex = vp9_quantizer_to_qindex(picture_control_set_ptr->picture_qp);
             }
-
+			if (encodeContextPtr->vbv_buf_size && encodeContextPtr->vbv_max_rate)
+			{
+				eb_block_on_mutex(encodeContextPtr->sc_buffer_mutex);
```
Comment (Austin-Hu) [inline Source/Lib/Codec/EbRateControlProcess.c]: To avoid race condition with SpeedBufferControl() function?

**Comment by karthikmcw** [comment]
> Hi @karthikmcw,
> Please help add Signed-off for your commits to pass DCO.
> And do you have any plan to add the "hrd" option like SVT-HEVC? Otherwise VBV verification couldn't be done via VQAnalyzer which requires HRD to check buffer fullness, and we've to use other video analyzers (such as Vega Media Analyzer) to verify it. Thanks!

HRD related SEI messages are not supported as per VP9 standard.

**Comment by karthikmcw** [comment]
> Hi @karthikmcw,
> Please help add Signed-off for your commits to pass DCO.
> And do you have any plan to add the "hrd" option like SVT-HEVC? Otherwise VBV verification couldn't be done via VQAnalyzer which requires HRD to check buffer fullness, and we've to use other video analyzers (such as Vega Media Analyzer) to verify it. Thanks!

HRD related SEI messages are not supported as per VP9 standard.

**Comment by Austin-Hu** [comment]
> > Hi @karthikmcw,
> > Please help add Signed-off for your commits to pass DCO.
> > And do you have any plan to add the "hrd" option like SVT-HEVC? Otherwise VBV verification couldn't be done via VQAnalyzer which requires HRD to check buffer fullness, and we've to use other video analyzers (such as Vega Media Analyzer) to verify it. Thanks!
> 
> HRD related SEI messages are not supported as per VP9 standard.

Hi @karthikmcw,

So how did you verify the overflow or underflow of CPB for VP9 bit stream, with different vbv_maxrate and vbv_bufsize settings? The video analyzer tools (such as VQAnalyzer or Vega Media Analyzer) can only analyze buffer fullness with HRD set in encoded video stream. Thanks!

**Comment by karthikmcw** [comment]
> Hi @karthikmcw,
> Please help add Signed-off for your commits to pass DCO.
> And do you have any plan to add the "hrd" option like SVT-HEVC? Otherwise VBV verification couldn't be done via VQAnalyzer which requires HRD to check buffer fullness, and we've to use other video analyzers (such as Vega Media Analyzer) to verify it. Thanks!
> 
> HRD related SEI messages are not supported as per VP9 standard.
> 
> Hi @karthikmcw,
> So how did you verify the overflow or underflow of CPB for VP9 bit stream, with different vbv_maxrate and vbv_bufsize settings? The video analyzer tools (such as VQAnalyzer or Vega Media Analyzer) can only analyze buffer fullness with HRD set in encoded video stream. Thanks!

We haved not verified the overflow and underflow of CPB for VP9 bitstream, since HRD related SEI messages could not be signalled.

**Comment by Austin-Hu** [comment]
> > Hi @karthikmcw,
> > Please help add Signed-off for your commits to pass DCO.
> > And do you have any plan to add the "hrd" option like SVT-HEVC? Otherwise VBV verification couldn't be done via VQAnalyzer which requires HRD to check buffer fullness, and we've to use other video analyzers (such as Vega Media Analyzer) to verify it. Thanks!
> > HRD related SEI messages are not supported as per VP9 standard.
> > Hi @karthikmcw,
> > So how did you verify the overflow or underflow of CPB for VP9 bit stream, with different vbv_maxrate and vbv_bufsize settings? The video analyzer tools (such as VQAnalyzer or Vega Media Analyzer) can only analyze buffer fullness with HRD set in encoded video stream. Thanks!
> 
> We haved not verified the overflow and underflow of CPB for VP9 bitstream, since HRD related SEI messages could not be signalled.

@kavithasampath & @tianjunwork , so we couldn't verify the overflow or underfor for VBV with VP9 encoded clips.

@kavithasampath , please help address my inline comments and questions, so that I would review again. Thanks!

**Comment by karthikmcw** [comment]
> Hi @karthikmcw,
> Please help add Signed-off for your commits to pass DCO.
> And do you have any plan to add the "hrd" option like SVT-HEVC? Otherwise VBV verification couldn't be done via VQAnalyzer which requires HRD to check buffer fullness, and we've to use other video analyzers (such as Vega Media Analyzer) to verify it. Thanks!

We have added Signed-off for the latest commit

File: Source/API/EbSvtVp9Enc.h [hunk 2]
```diff
@@ -352,10 +352,22 @@ typedef struct EbSvtVp9EncConfiguration
     uint32_t                 recon_file;
 
 
-
     // Input stride
     uint32_t               input_picture_stride; // Includes padding
```
[the remark this answers is not in the document]
Comment (karthikmcw) [reply Source/API/EbSvtVp9Enc.h]: Our latest commit fixes the indentations

File: Source/Lib/Codec/EbPacketizationProcess.c [hunk 3]
```diff
@@ -628,7 +629,19 @@ void* packetization_kernel(void *input_ptr)
...
+			/* update VBV plan */
+			if (encode_context_ptr->vbv_max_rate && encode_context_ptr->vbv_buf_size)
+			{
+				int64_t buffer_fill_temp = (int64_t)(encode_context_ptr->buffer_fill);
+
+				buffer_fill_temp -= queue_entry_ptr->actualBits;
+				buffer_fill_temp = MAX(buffer_fill_temp, 0);
+				buffer_fill_temp = (int64_t)(buffer_fill_temp + (encode_context_ptr->vbv_max_rate * (1.0 / (sequence_control_set_ptr->frame_rate >> RC_PRECISION))));
+				buffer_fill_temp = MIN(buffer_fill_temp, encode_context_ptr->vbv_buf_size);
+				encode_context_ptr->buffer_fill = (uint64_t)(buffer_fill_temp);
```
[the remark this answers is not in the document]
Comment (karthikmcw) [reply Source/Lib/Codec/EbPacketizationProcess.c]: Yes, we shall remove this.

**Comment by karthikmcw** [comment]
> Hi @karthikmcw,
> Please help add Signed-off for your commits to pass DCO.
> And do you have any plan to add the "hrd" option like SVT-HEVC? Otherwise VBV verification couldn't be done via VQAnalyzer which requires HRD to check buffer fullness, and we've to use other video analyzers (such as Vega Media Analyzer) to verify it. Thanks!
> 
> We have added Signed-off for the latest commit
> 
> Your Signed-off hasn't been added correctly with your latest commit, and you can refer to your whole change here.

Apparently, a signed-off message ( Signed-off-by: karthik <karthik@multicorewareinc.com> ) is added as part of the commit message. Do let us know if something more has to be done on this regard.

File: Source/Lib/Codec/EbPacketizationProcess.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (karthikmcw) [reply Source/Lib/Codec/EbPacketizationProcess.c]: This is done in our latest commit :)

File: Source/API/EbSvtVp9Enc.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (karthikmcw) [reply Source/API/EbSvtVp9Enc.h]: Yes, done in our latest commit.

File: Source/Lib/Codec/EbRateControlProcess.c [hunk 4]
```diff
@@ -4037,10 +4037,152 @@ void init_rc(
...
+		queue_entry_index_head_temp - HIGH_LEVEL_RATE_CONTROL_HISTOGRAM_QUEUE_MAX_DEPTH :
+		queue_entry_index_head_temp;
+
+	queue_entry_index_temp = queue_entry_index_head_temp;
+	bitrate_flag = encode_context_ptr->vbv_max_rate <= encode_context_ptr->available_target_bitrate;
+	int32_t current_ind = (queue_entry_index_temp > HIGH_LEVEL_RATE_CONTROL_HISTOGRAM_QUEUE_MAX_DEPTH - 1) ? queue_entry_index_temp - HIGH_LEVEL_RATE_CONTROL_HISTOGRAM_QUEUE_MAX_DEPTH : queue_entry_index_temp;
+
+
+	/* Avoid an infinite loop. */
+	for (int32_t iterations = 0; iterations < 1000 && loop_terminate != 3; iterations++)
```
[the remark this answers is not in the document]
Comment (karthikmcw) [reply Source/Lib/Codec/EbRateControlProcess.c]: Over each of the 1000 iterations, frame VBV algorithm increments/decrements the QP value in order to adjust the estimated buffer level to be between the extremes of overflow and underflow. It turns out that 1000 is a sufficiently large number to accomplish this in practice, and fine tuning the QP will be reached (and loop will be quit) well within these many iterations. However, if by any chance, this does not happen in 1000 iterations then it means that there is further scope for fine tuning the QP.

[the remark this answers is not in the document]
Comment (Austin-Hu) [reply Source/Lib/Codec/EbRateControlProcess.c]: Makes sense to me, thanks, @karthikmcw !

**Comment by hassount** [comment]
Tested the performance of this PR and it is deemed to be acceptable and similar to that of the VBV of SVT-HEVC

**Comment by intelmark** [comment]
Ignoring DCO and merging

**Pull request merged**